### PR TITLE
Winter cleaning

### PR DESCRIPTION
**Pull request by vhavlena**
A bit of winter cleaning. In particular:
- adding comments; removing obsolete stuff 
- fixing a small bug in expression checking
- getting rid of warnings during a build on linux

### CODE REVIEW

**Comment by vhavlena** [comment]
```
# of formulae: 103405
--------------------------------------------------
tool                                    ✅    ❌      time    avg    med    std    sat    unsat    unknown    TO    MO+ERR    other
----------------------------------  ------  ----  --------  -----  -----  -----  -----  -------  ---------  ----  --------  -------
z3-noodler-c46085b-56780ef          102593   812   9547.38   0.09   0.01   1.47  63082    39511        548   189        75        0
z3-noodler-da8bd67-56780ef          102593   812   9289.80   0.09   0.01   1.42  63083    39510        548   188        76        0
cvc5-1.2.0                           99841  3564  73861.19   0.74   0.01   6.42  60931    38910          2  3553         9        0
--------------------------------------------------
```

image

File: .gitignore [hunk 1]
```diff
@@ -100,3 +100,4 @@ CMakeSettings.json
 .DS_Store
 dbg/**
 *.wsp
+test-input
```
Comment (jurajsic) [inline .gitignore:103]:
```suggestion
test-input

```

File: src/smt/theory_str_noodler/expr_cases.cpp [hunk 2]
```diff
@@ -87,17 +87,10 @@ bool is_len_num_eq(expr* e, ast_manager& m, seq_util& m_util_s, arith_util& m_ut
...
-            return true;
-        }
         if(m_util_a.is_numeral(right, num) && m_util_s.str.is_length(left, len_arg)) {
             return true;
         }
     } else if(m.is_not(e, e_not) && m_util_a.is_ge(e_not, left, right)) {
-        if(m_util_a.is_numeral(left, num) && m_util_s.str.is_length(right, len_arg)) {
-            num--;
-            return true;
-        }
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/expr_cases.cpp:100]: Why was this removed?

File: src/smt/theory_str_noodler/expr_cases.h [hunk 3]
```diff
@@ -74,10 +71,45 @@ bool is_indexof_add(expr* e, expr* index_str, ast_manager& m, seq_util& m_util_s
  */
 bool is_indexof_at(expr * index_param, expr* index_str, ast_manager& m, seq_util& m_util_s);
 
+/**
+ * @brief Check if the expression @p e is of the form to_int(x) = num where num 
+ * is a number (or num = to_int(x)).
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/expr_cases.h:76]:
```suggestion
 * @brief Check if the expression @p e is of the form to_int(x) = num (or num = to_int(x)) where num is a number .
```
For `@brief`, it needs to be on one line I think..

File: src/smt/theory_str_noodler/expr_cases.h [hunk 4]
```diff
@@ -74,10 +71,45 @@ bool is_indexof_add(expr* e, expr* index_str, ast_manager& m, seq_util& m_util_s
...
+ * @param m_util_a arith ast util
+ * @param[out] to_int_arg Argument of to_int (x)
+ * @param[out] num Number on the opposite side
+ * @return true <-> if of the particular form.
+ */
 bool is_to_int_num_eq(expr* e, ast_manager& m, seq_util& m_util_s, arith_util& m_util_a, expr*& to_int_arg, rational& num);
 
+/**
+ * @brief Check if the expression @p e is of the form len(x) = num where num 
+ * is a number (or num = len(x)).
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/expr_cases.h:90]:
```suggestion
 * @brief Check if the expression @p e is of the form len(x) = num (or num = len(x)) where num is a number.
```

File: src/smt/theory_str_noodler/expr_cases.h [hunk 5]
```diff
@@ -74,10 +71,45 @@ bool is_indexof_add(expr* e, expr* index_str, ast_manager& m, seq_util& m_util_s
...
 bool is_indexof_at(expr * index_param, expr* index_str, ast_manager& m, seq_util& m_util_s);
 
+/**
+ * @brief Check if the expression @p e is of the form to_int(x) = num where num 
+ * is a number (or num = to_int(x)).
+ * 
+ * @param e Expression to be checked
+ * @param m Ast manager
+ * @param m_util_s string ast util
+ * @param m_util_a arith ast util
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/expr_cases.h:81]: I think generally we should skip these comments that do not say anything, you get the same information from the types of the arguments (I would remove this from other functions descriptions too that were added in this PR).

File: src/smt/theory_str_noodler/formula_preprocess.cpp [hunk 6]
```diff
@@ -629,7 +629,7 @@ namespace smt::noodler {
             Concat side = occurrs.begin()->position > 0 ? act_pred.get_right_side() : act_pred.get_left_side();
 
             int start = std::abs(occurrs.begin()->position) - 1;
-            for(int i = start; i < side.size(); i++) {
+            for(int i = start; i < (int)side.size(); i++) {
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/formula_preprocess.cpp:632]: Why not
```suggestion
            for (size_t i = start; i < side.size(); i++) {
```
?
Does `side.size()` return signed?

File: src/smt/theory_str_noodler/theory_str_noodler.cpp [hunk 7]
```diff
@@ -489,27 +461,14 @@ namespace smt::noodler {
...
-            // } else {
-            //     util::throw_error("unsupported predicate");
-            //     // handle_not_suffix(e);
-            // }
+            // already handled in relevant_eh. It suffices to handle is_suffix only once as it is fully 
+            // axiomatized using basic string constraints ((dis)equations, regexes, and lengths)
         } else if (m_util_s.str.is_contains(e, e1, e2)) {
-            // FIXME could the problem from previous two also occur here? if yes, handle_contains and handle_not_contains should be in relevant_eh BUT m_not_contains_todo should be updated probably here (if applicable), so we can return unknown
+            // notcontains cannot be fully axiomatized. We need to add it among string constraints solved later in final_check 
+            // (it is not sufficient to add it only once at the beginning of the solver run as it max occurr in different SAT assignments).
```
Comment (jurajsic) [inline src/smt/theory_str_noodler/theory_str_noodler.cpp:471]: For me this still does not explain why there is no problem with handling `contains` here, but there is problem for the previous ones. Generally speking, I think we should move towards handling everything here and not in `relevant` as we do now, or at least Z3 does it this way.

**Review by jurajsic — APPROVED** [review]
Looks good, just some thoughts.

File: .gitignore (hunk 1 above)
Comment (vhavlena) [reply .gitignore:103]: I put it there deliberately. I am using the `test-input` directory to collect formulae (e.g., those that we gave a wrong result at some point).

File: src/smt/theory_str_noodler/expr_cases.h (hunk 5 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/expr_cases.h:81]: It is fine to keep it there for the IDE. Moreover if a parameter comment is missing it might look like a mistake (doxygen would probably complain too). I am more thinking of some pattern class instead of just pure functions. In that case these common parameters could be stored as member variables.

File: src/smt/theory_str_noodler/formula_preprocess.cpp (hunk 6 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/formula_preprocess.cpp:632]: I was not sure if `start` might be negative or not. `side.size()` should be `size_t` imo.

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 7 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:471]: I think it could be safe to move the handling of `contains` to the relevant only (It needs to be checked on the benchmark, I haven't tried, so I left it untouched). In general it would be nice to have it in `assign_eh` but it requires storing information about which predicates were already axiomatized at which decision levels (the `scoped_vector` datastructure is probably fine for that but you need to somehow properly integrate to the noodler's internals)..

File: src/smt/theory_str_noodler/expr_cases.cpp (hunk 2 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/expr_cases.cpp:100]: I realized that it is wrong (probably some kind of copy-paste error). It basically didn't distinguished between `a <= 5` and `5 <= a` (the code that uses this function assumes only one case).

File: .gitignore (hunk 1 above)
Comment (jurajsic) [reply .gitignore:103]: It was about missing newline at the end of file 😄

File: src/smt/theory_str_noodler/expr_cases.cpp (hunk 2 above)
Comment (jurajsic) [reply src/smt/theory_str_noodler/expr_cases.cpp:100]: Ah ok, good catch.

File: src/smt/theory_str_noodler/expr_cases.h (hunk 5 above)
Comment (jurajsic) [reply src/smt/theory_str_noodler/expr_cases.h:80]: Hmm, you're right. I sometimes wonder if we shouldn't just put everything into `theory_str_noodler` class at this point (but keep it in separate files).

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 7 above)
Comment (jurajsic) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:471]: But we do not do the storing about this information currently for `contains` right? So if we keep it here for now, it could lead to bugs, or am I wrong?

File: src/smt/theory_str_noodler/formula_preprocess.cpp (hunk 6 above)
Comment (jurajsic) [reply src/smt/theory_str_noodler/formula_preprocess.cpp:632]: Ah, I see it now. Hmm, but then the code inside the loop could be buggy, we are doing `side[i]` a lot.

Do you remember why `start` is initialized the way it is initalized? Why is there `std::abs(occurrs.begin()->position)`? Can `occurrs.begin()->position` be even negative?

Comment (vhavlena) [reply src/smt/theory_str_noodler/formula_preprocess.cpp]: I have checked it. It is supposed to be positive. Fixed it.

File: src/smt/theory_str_noodler/theory_str_noodler.cpp (hunk 7 above)
Comment (vhavlena) [reply src/smt/theory_str_noodler/theory_str_noodler.cpp:471]: Currently the `handle_contains` is redundant there because it is handled already in `relevant_eh` (since we cache the occurrences, it does nothing in `assign_eh`). I removed it and it works ok (no wrong results nor performance degradation).

**Comment by jurajsic** [comment]
I think you can merge.
